pixeltable/share/packager.py

```diff
@@ -104,86 +110,116 @@ def __export_table(self, t: catalog.Table) -> None:
-        # The data is first loaded from the DataFrame into a sequence of pyarrow tables, batched in order to avoid
-        # excessive memory usage. The pyarrow tables are then amalgamated into the (single) Iceberg table on disk.
-        for pa_table in self.__to_pa_tables(df, actual_col_types, iceberg_schema):
-            iceberg_tbl.append(pa_table)
+        # Add columns for the primary key components of the base table rows.
+        # We need to use a VminRef for the vmin component, to ensure we get the correct vmin for the base table
+        # (which may be different from the vmin of the primary table).
+        # We also explicitly select the rowid components, in order to use them with the group_by() / any_value()
+        # pattern to deduplicate in SQL (see below).
+        rowid_len = len(tv.store_tbl.rowid_columns())
+        for idx in range(rowid_len):
+            select_exprs[f'pk_{idx}'] = exprs.RowidRef(tvp.tbl_version, idx)
+        select_exprs['pk_vmin'] = exprs.VminRef(tvp.tbl_version)
 
-    @classmethod
-    def __iceberg_namespace(cls, table: catalog.Table) -> str:
-        """
-        Iceberg tables must have a namespace, which cannot be the empty string, so we prepend `pxt` to the table path.
-        """
-        parent_path = table._parent()._path()
-        if len(parent_path) == 0:
-            return 'pxt'
+        if tv.id == self.table._tbl_version.id:
+            # Selecting from the primary table: just a simple select statement
+            df = self.table.select(**select_exprs)
         else:
-            return f'pxt.{parent_path}'
+            # Selecting from an ancestor table: in this case we still need to run the select() with `self.table` as the
+            # context, not the base TableVersionPath, so that we export only those rows that are actually present in
+            # `self.table`. The group_by() is needed to handle the case of iterator views correctly; it ensures that
+            # records of the base table are appropriately deduplicated on rowid.
+            select_exprs = {name: pxtf.any_value(expr) for name, expr in select_exprs.items()}
+            df = self.table._df().group_by(tv).select(**select_exprs)
+
+        parquet_schema = self.__to_parquet_schema(df._schema)
+        # The parquet file naming scheme anticipates future support for partitioning.
+        parquet_dir = self.tables_dir / f'tbl_{tv.id.hex}'
+        parquet_dir.mkdir()
+        parquet_file = parquet_dir / f'tbl_{tv.id.hex}.00000.parquet'
+        _logger.info(f'Creating parquet table: {parquet_file}')
+
+        # Populate the Parquet table with data.
+        # The data is first loaded from the DataFrame into a sequence of pyarrow tables, batched in order to avoid
+        # excessive memory usage. The pyarrow tables are then amalgamated into the (single) Parquet table on disk.
+        # We use snappy compression for the Parquet tables; the entire bundle will be bzip2-compressed later, so
```

Comment:
Did you check whether bzip2 makes a difference? Parquet already has various forms of compression baked in.

Reply:
The tarball contains content other than parquet files, so I'd presume we need to compress it at the end regardless? (For example, it might contain compressible pdf or html documents.)